server/src/main/java/org/opensearch/index/store/FsDirectoryFactory.java

```diff
@@ -169,7 +174,7 @@ public IndexInput openInput(String name, IOContext context) throws IOException {
 
         boolean useDelegate(String name) {
             final String extension = FileSwitchDirectory.getExtension(name);
-            return mmapExtensions.contains(extension);
+            return !nioExtensions.contains(extension);
```

Comment:
Yes, you are correct about user supplying empty [] nio list will not do as expected. I am not sure what is the recommended path to avoid this case. Is there a better way to detect when user defines `index.store.hybrid.nio.extensions`? Or make default value something other than `Collections.emptyList()`?

Reply:
Would this work as expected if the default value of `index.store.hybrid.nio.extensions` was defined as:

```
"segments_N", "write.lock", "si", "cfe", "fnm", "fdx", "fdt", "pos", "pay", "nvm", "dvm", "tvx", "tvd", "liv", "dii", "vec", "vem"
```

I think the following logic already implicitly implements this:

```
if (nioExtensions.isEmpty()) {
    nioExtensions.addAll(INDEX_STORE_HYBRID_ALL_EXTENSIONS);
    nioExtensions.removeAll(indexSettings.getValue(IndexModule.INDEX_STORE_HYBRID_MMAP_EXTENSIONS));
}
```

but defining the above list as the explicit default might be simpler.

Reply:
@andrross - David and I discussed this issue yesterday, and here is what I think:

* We should think of these two settings (mmap/niofs) as complementary. Setting any one updates the both of them with the difference from universal set. That way the lists are always consistent and there is no confusion to the users
* Eventually in 4.0, we will remove the mmap setting so only nio one remains
* Although if Lucene adds new file format in the interim (not part of above extension list), the behavior is less predictable. Depending on the value, it might use mmap or nio for that new extension

Other way to do is:

* Throw an error whenever user tries to update mmap setting saying "Use niofs setting instead of mmap"
* In 4.0, we will remove the setting itself so it throws unrecognized setting
* Only during upgrades if mmap setting was added earlier, add the complement set as nio setting

Reply:
@jainankitk 

I think the crux of this issue is that (as far as I know), classes that use the `Settings` API cannot differentiate between a value being unset and a value being set to the default value. So assuming an empty list is the default isn't actually valid in the current implementation, because it gives odd behavior if the user explicitly sets the value to be an empty list. I think we can work around it with the following logic:

- The default of the `mmap.extensions` setting is the list of mmap extensions
- The default of the `nio.extensions` setting is the list of nio extensions
- The union of the two defaults is the set of all Lucene file extensions that exist today ("universal set")
- The settings validation logic ensures that only one of the two settings is explicitly set by the user
- The logic for the hybrid fs in FSDirectory is as follows:
  - if the `mmap.extensions` setting is not the default value, then compute the `nioExtensions` list by doing `universal set - mmap.extensions settings value`
  - else if the `mmap` setting is the default value, then the `nioExtenions` list can just be the value of the `nio.extensions` setting

Does this work?

Reply:
Yes, the main challenge is differentiating the default setting from user-specified value. I think @andrross's logic will work. I will also include @jainankitk's recommendation to add validation on both mmap and nio sides.

Reply:
@andrross - I like your suggestion, though one minor catch for new lucene extension. From my understanding of above logic, if user specified `mmap.extensions` will the new extension use mmapfs? Ideally, it should use niofs as the user explicitly provided the list of `mmap.extensions`. I am even okay with this nuance as `mmap.extensions` is on deprecation path, and we want to encourage users to use `niofs.extensions`, just want to call out explicitly

Reply:
If a new extension is added, it will always default to mmap. Even if user omits it from `index.store.hybrid.mmap.extensions`. This is the behavior we wanted in the original github issue with a small caveat since we are maintaining both settings. The only way to load a NEW file type with nio is by including it in `index.store.hybrid.nio.extensions`.